src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi

```diff
@@ -254,34 +272,81 @@ def call_credentials_metadata_plugin(CredentialsMetadataPlugin plugin):
-  for i in range(credentials.c_ssl_pem_key_cert_pairs_count):
-    credentials.c_ssl_pem_key_cert_pairs[i] = (
-        (<SslPemKeyCertPair>pem_key_cert_pairs[i]).c_pair)
-  credentials.c_credentials = grpc_ssl_server_credentials_create(
-      c_pem_root_certs, credentials.c_ssl_pem_key_cert_pairs,
-      credentials.c_ssl_pem_key_cert_pairs_count,
-      GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY if force_client_auth else GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE,
-      NULL)
+  credentials.c_ssl_pem_key_cert_pairs = _create_c_ssl_pem_key_cert_pairs(pem_key_cert_pairs)
+  cdef grpc_ssl_server_certificate_config *c_cert_config = NULL
+  c_cert_config = grpc_ssl_server_certificate_config_create(
+    c_pem_root_certs, credentials.c_ssl_pem_key_cert_pairs,
+    credentials.c_ssl_pem_key_cert_pairs_count)
+  cdef grpc_ssl_server_credentials_options* c_options = NULL
+  # C-core assumes ownership of c_cert_config
+  c_options = grpc_ssl_server_credentials_create_options_using_config(
+    GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY if force_client_auth else GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE,
+    c_cert_config)
+  # C-core assumes ownership of c_options
+  credentials.c_credentials = grpc_ssl_server_credentials_create_with_options(c_options)
   return credentials
 
+def server_certificate_config_ssl(pem_root_certs, pem_key_cert_pairs):
+  pem_root_certs = str_to_bytes(pem_root_certs)
+  pem_key_cert_pairs = list(pem_key_cert_pairs)
+  cdef ServerCertificateConfig cert_config = ServerCertificateConfig()
+  cert_config.references.append(pem_root_certs)
+  cert_config.references.append(pem_key_cert_pairs)
+  cert_config.c_pem_root_certs = _get_c_pem_root_certs(pem_root_certs)
+  cert_config.c_ssl_pem_key_cert_pairs_count = len(pem_key_cert_pairs)
+  cert_config.c_ssl_pem_key_cert_pairs = _create_c_ssl_pem_key_cert_pairs(pem_key_cert_pairs)
+  cert_config.c_cert_config = grpc_ssl_server_certificate_config_create(
+    cert_config.c_pem_root_certs, cert_config.c_ssl_pem_key_cert_pairs,
+    cert_config.c_ssl_pem_key_cert_pairs_count)
+  return cert_config
+
+def server_credentials_ssl_dynamic_cert_config(initial_cert_config,
+                                               cert_config_fetcher,
+                                               bint force_client_auth):
+  if initial_cert_config is None:
+    raise ValueError('initial certificate config must be specified')
+  if not callable(cert_config_fetcher):
+    raise ValueError('certificate config fetcher must be callable')
+  cdef ServerCredentials credentials = ServerCredentials()
+  credentials.initial_cert_config = initial_cert_config
+  credentials.cert_config_fetcher = cert_config_fetcher
+  cdef grpc_ssl_server_credentials_options* c_options = NULL
+  c_options = grpc_ssl_server_credentials_create_options_using_config_fetcher(
+    GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY if force_client_auth else GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE,
```

Comment:
This large conditional expression looks like it should be spread across three lines? We don't have a linter running on Cython code but we do try to keep it limited to seventy-nine columns wide.

Reply:
what's line-breaking convention used in grpc python? three lines like this?
1.
```
GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY \
if force_client_auth \
else GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE,
```
or 2.
```
  cdef grpc_ssl_client_certificate_request_type client_certificate_request = \
      GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY
  if not force_client_auth:
    client_certificate_request = GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE
  c_options = grpc_ssl_server_credentials_create_options_using_config_fetcher(
    client_certificate_request,
```
or 3.
```
  cdef grpc_ssl_client_certificate_request_type client_certificate_request
  if force_client_auth:
    client_certificate_request = CRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY
  else:
    client_certificate_request = GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE
  c_options = grpc_ssl_server_credentials_create_options_using_config_fetcher(
    client_certificate_request,
```
thanks!

Reply:
Not 1, because [we never use backslash](https://google.github.io/styleguide/pyguide.html?showone=Line_length#Line_length). Not 2, because we avoid assigning a value to a field until after determining that the value assigned to the field is the right value to assign to the field. 3 looks okay, but I was thinking more along the lines of 4:
```py
c_options = grpc_ssl_server_credentials_create_options_using_config_fetcher(
    GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY
    if force_client_auth else
    GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE,
    c_cert_config)
```
. We almost always keep to the rule "if a conditional expression doesn't fit on one line, refactor it into a proper `if/else`", but I think this is a reasonable place to use the embedding of the conditional expression within the function call to avoid backslash continuation or a local field.

3 is okay too.

Reply:
thanks! fixed